www/source/blog/2018-08-18-lift-and-shift-hab.html.md

```diff
@@ -0,0 +1,396 @@
+
+## Running Table Setting locally
+
+Let's explore how to run Table Setting locally before getting started with Habitat.  As a prerequisite, Python 3.7 should be installed.  In fact Python 3.* should work but no promises!  This quickly brings us to the downside of shipping anything other than a single, executable artifact as disparity can easily creep in!  Quietly skipping over the potential trauma of ensuring we have the correct Python in our `PATH`, Table Setting can be run via:
+
+````bash
+$ git clone https://github.com/skpaterson/table-setting
+$ cd table-setting
+$ python -m venv tsenv
+$ . tsenv/bin/activate
+$ pip install -r requirements.txt
+$ python run_app.py 
+````
+
+Navigating to [http://localhost:5000](http://localhost:5000) you should see the Table Setting Swagger UI:
+
+![Running Local InMemory SQLite](media/2018-08-18-lift-and-shift-hab/local_in_memory.png)
+
+Feel free to try creating some forks!  By default this runs with a clean SQLite in-memory database that is wiped on restart and fork CRUD operations are available. 
+
+A user running even this simple application locally is perhaps unnecessarily exposed to several things:
+* Source control with `git`
+* Python language installation
+* Virtual environment setup
+* Python packaging with `pip` 
+* Internal application structure e.g. where to find the thing to run
+
+A similar list could likely be formulated for applications written in other languages.  In any case, this complicates life for performing any changes at a later stage e.g.
+* Switching to a MySQL database backend
+* Running on a new machine, or
+* Moving to a (new) Cloud provider or service
+
+
+## Packaging Table Setting with Habitat
+
+The following directory structure within the Table Setting repository relates to Habitat packaging:
+
+```bash
+$ tree habitat/
+habitat/
+├── README.md
+├── config
+├── default.toml
+├── hooks
+│   ├── init
+│   └── run
+└── plan.sh
+```
+This incorporates a set of steps to install required dependencies, similar to those taken when running locally and also includes the MySQL library setup.  Table Setting explicitly depends on a known good Python version. Environment variables are used to convey runtime configuration to Table Setting, some are conditionally set when e.g. binding to a Habitat managed MySQL database.  
+
+Let's concentrate for now on creating deployment artifacts with Habitat.  Below we use the Studio to export different package formats:
+```bash
+$ cd table-setting
+$ hab studio enter
+[STUDIO] build
+[STUDIO] hab pkg export tar $(ls -1t results/*.hart | head -1)
+[STUDIO] hab pkg export docker $(ls -1t results/*.hart | head -1)
+```
```

Comment:
Instead of doing this `ls` trick, I recommend sourcing the `results/last_build.env` file and making use of the environment variables in there.

So:
```bash
build
source results/last_build.env
hab pkg export tar "results/${pkg_artifact}"
hab pkg export docker "results/${pkg_artifact}"
```

Reply:
Thanks, I didn't know about this one, is it documented somewhere?

Reply:
It looks like it has been mentioned in passing in a handful of blog posts, and also [here](https://www.habitat.sh/docs/developing-packages/#non-interactive-build), but I don't know that we've ever explicitly highlighted it as a best practice.

I'll see about getting this more clearly documented, as well.